sleepecg/feature_extraction.py

```diff
@@ -522,65 +653,25 @@ def extract_features(
-    rri_required = 'hrv-time' in required_groups or 'hrv-frequency' in required_groups
-
-    features = []
-    stages = []
 
-    for record in records:
-        if record.sleep_stages is not None and record.sleep_stage_duration is not None:
-            record_duration = (len(record.sleep_stages) - 1) * record.sleep_stage_duration
-        elif record.heartbeat_times is not None:
-            record_duration = record.heartbeat_times[-1]
-        else:
-            raise ValueError(f"Record duration can't be inferred for {record.id}. ")
-        num_stages = record_duration // sleep_stage_duration
-        stage_times = np.arange(num_stages) * sleep_stage_duration
-
-        if rri_required:
-            if record.heartbeat_times is None:
-                raise ValueError(f"Can't extract HRV features for record {record.id} without heartbeat_times.")  # noqa: E501
-            rri = preprocess_rri(
-                np.diff(record.heartbeat_times),
-                min_rri=min_rri,
-                max_rri=max_rri,
-            )
-            rri_times = record.heartbeat_times[1:]
-
-        X = []
-        for feature_group in required_groups:
-            if feature_group == 'hrv-time':
-                X.append(
-                    _hrv_timedomain_features(
-                        rri,
-                        rri_times,
-                        stage_times,
-                        lookback,
-                        lookforward,
-                    ),
-                )
-            elif feature_group == 'hrv-frequency':
-                X.append(
-                    _hrv_frequencydomain_features(
-                        rri,
-                        rri_times,
-                        stage_times,
-                        lookback,
-                        lookforward,
-                        fs_rri_resample,
-                        max_nans,
-                        feature_ids,
-                    ),
-                )
-        features.append(np.hstack(X)[:, col_indices])
-        if record.sleep_stages is None or sleep_stage_duration == record.sleep_stage_duration:  # noqa: E501
-            stages.append(record.sleep_stages)
-        else:
-            interp_stages = interp1d(
-                np.arange(len(record.sleep_stages)) * record.sleep_stage_duration,
-                record.sleep_stages,
-                kind='nearest',
-            )(stage_times)
-            stages.append(interp_stages)
+    # _extract_features_single has two return values, so the list returned
+    # by _parallel needs to be unpacked
+    Xy = _parallel(
+        n_jobs,
+        _extract_features_single,
+        records,
+        sleep_stage_duration,
+        min_rri,
+        max_rri,
+        required_groups,
+        lookback,
+        lookforward,
+        fs_rri_resample,
+        max_nans,
+        feature_ids,
+        col_indices,
+    )
+    features = [X for X, _ in Xy]
+    stages = [y for _, y in Xy]
```

Comment:
Would that work too?

```suggestion
    features, stages = zip(*Xy)
```

Reply:
yes, but the types wouldn't propagate correctly :/

Reply:
Did I mention I really don't like type annotations in Python?

Reply:
once or twice :D